servers/services/src/main/java/org/projectnessie/services/impl/NamespaceApiImpl.java

```diff
@@ -0,0 +1,299 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.services.impl;
+
+import com.google.common.collect.Sets;
+import java.security.Principal;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.projectnessie.api.NamespaceApi;
+import org.projectnessie.api.params.MultipleNamespacesParams;
+import org.projectnessie.api.params.NamespaceParams;
+import org.projectnessie.error.NessieNamespaceAlreadyExistsException;
+import org.projectnessie.error.NessieNamespaceNotEmptyException;
+import org.projectnessie.error.NessieNamespaceNotFoundException;
+import org.projectnessie.error.NessieReferenceNotFoundException;
+import org.projectnessie.model.CommitMeta;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.Content.Type;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.GetNamespacesResponse;
+import org.projectnessie.model.ImmutableGetNamespacesResponse;
+import org.projectnessie.model.Namespace;
+import org.projectnessie.model.Operation.Delete;
+import org.projectnessie.model.Operation.Put;
+import org.projectnessie.services.authz.Authorizer;
+import org.projectnessie.services.config.ServerConfig;
+import org.projectnessie.versioned.BranchName;
+import org.projectnessie.versioned.Key;
+import org.projectnessie.versioned.Operation;
+import org.projectnessie.versioned.ReferenceConflictException;
+import org.projectnessie.versioned.ReferenceNotFoundException;
+import org.projectnessie.versioned.VersionStore;
+import org.projectnessie.versioned.WithType;
+
+public class NamespaceApiImpl extends BaseApiImpl implements NamespaceApi {
+
+  public NamespaceApiImpl(
+      ServerConfig config,
+      VersionStore<Content, CommitMeta, Content.Type> store,
+      Authorizer authorizer,
+      Principal principal) {
+    super(config, store, authorizer, principal);
+  }
+
+  @Override
+  public Namespace createNamespace(NamespaceParams params)
+      throws NessieNamespaceAlreadyExistsException, NessieReferenceNotFoundException {
+    try {
+      BranchName branch = branchFromRefName(params.getRefName());
+
+      Callable<Void> validator =
+          () -> {
+            if (getExplicitlyCreatedNamespace(params, branch).isPresent()) {
+              throw namespaceAlreadyExistsException(params);
```

Comment:
Technically, it could be a table with the same key, right? Should we adjust the message to say "Another content object already exists"... or something like that?

Reply:
ok, let's address in a follow-up PR

Reply:
done in https://github.com/projectnessie/nessie/pull/3636